### PR TITLE
Handle method parameters with defaults correctly.

### PR DESCRIPTION
**Pull request by aomarks**
Also:
- Add a "custom" test fixture to manually add coverage for specific cases.
- Output rest method parameter even when it has no JSDoc type.

### CODE REVIEW

**Comment by aomarks** [comment]
Retracting to work on this PR a little more.

**Comment by aomarks** [comment]
Updated this PR, ready for review!

It may be helpful to view the 3 commits separately.

File: src/gen-ts.ts [hunk 1]
```diff
@@ -297,9 +299,34 @@ function handleMethods(analyzerMethods: Iterable<analyzer.Method>):
     });
     m.description = method.description || '';
 
-    for (const param of method.params || []) {
-      const {type, optional, rest} = closureParamToTypeScript(param.type);
-      m.params.push(new ts.Param({name: param.name, type, optional, rest}));
+    let requiredAhead = false;
+    for (let i = (method.params || []).length - 1; i >= 0; i--) {
```
Comment (rictic) [inline src/gen-ts.ts]: Arg, complicated three arg for loop. What about:

```typescript
function* reversedIter<T>(arr: T[]) {
  for (let i = arr.length - 1; i >= 0; i--) {
    yield arr[i];
  }
}
```

Then iteration can be:

```typescript
for (const param of reversed(method.params || [])) {

}
```

File: src/gen-ts.ts [hunk 2]
```diff
@@ -297,9 +299,34 @@ function handleMethods(analyzerMethods: Iterable<analyzer.Method>):
...
+        }
+      } else if (!optional) {
+        requiredAhead = true;
+      }
+
+      // Analyzer might know this is a rest parameter even if there was no
+      // JSDoc type annotation (or if it was wrong).
+      rest = rest || !!param.rest;
+      if (rest && type.kind !== 'array') {
+        // Rest types are always arrays.
```
Comment (rictic) [inline src/gen-ts.ts]: Maybe worth adding that the closure type is written without the Array syntax

**Review by rictic — APPROVED** [review]
Nice! Subtle change

File: src/gen-ts.ts (hunk 1 above)
Comment (aomarks) [reply src/gen-ts.ts]: Done.

File: src/gen-ts.ts (hunk 2 above)
Comment (aomarks) [reply src/gen-ts.ts]: Done.

**Pull request merged**